web/client-api/src/main/java/io/deephaven/web/client/api/barrage/stream/BiDiStream.java

```diff
@@ -0,0 +1,226 @@
+    public abstract void onStatus(JsConsumer<ResponseStreamWrapper.Status> handler);
+    public abstract void onEnd(JsConsumer<ResponseStreamWrapper.Status> handler);
+
+    static class WebsocketBiDiStream<T, U> extends BiDiStream<T, U> {
+        @JsType(isNative = true, name = "Object", namespace = JsPackage.GLOBAL)
+        private static class BidirectionalStreamWrapper<ReqT, ResT> {
+            native void cancel();
+
+            native void end();
+
+            native WebsocketBiDiStream<ReqT, ResT> on(String type, Function handler);
+
+            native WebsocketBiDiStream<ReqT, ResT> write(ReqT message);
+        }
+        private final BidirectionalStreamWrapper<T, U> wrapped;
+
+        WebsocketBiDiStream(BidirectionalStreamWrapper<T, U> wrapped) {
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public void send(T payload) {
+            wrapped.write(payload);
+        }
+
+        @Override
+        public void cancel() {
+            wrapped.cancel();
+        }
+        @Override
+        public void end() {
+            wrapped.end();
+        }
+
+        @Override
+        public void onData(JsConsumer<U> handler) {
+            wrapped.on("data", Js.cast(handler));
+        }
+        @Override
+        public void onStatus(JsConsumer<ResponseStreamWrapper.Status> handler) {
+            wrapped.on("status", Js.cast(handler));
+        }
+        @Override
+        public void onEnd(JsConsumer<ResponseStreamWrapper.Status> handler) {
+            wrapped.on("end", Js.cast(handler));
+        }
+    }
+
+    static class EmulatedBiDiStream<T, U> extends BiDiStream<T, U> {
+        private final JsFunction<T, ResponseStreamWrapper<U>> responseStreamFactory;
+        private final JsArray<JsConsumer<ResponseStreamWrapper<U>>> pending = new JsArray<>();
+        private final int intTicket;
+
+        private ResponseStreamWrapper<U> responseStream;
+        private final NextStreamMessageFactory<T> nextWrapper;
+        private final Supplier<BrowserHeaders> headers;
+
+        private int nextSeq = 0;
+
+        EmulatedBiDiStream(OpenStreamFactory<T> responseStreamFactory, NextStreamMessageFactory<T> nextWrapper, int intTicket, Supplier<BrowserHeaders> headers) {
+            this.responseStreamFactory = firstReq -> ResponseStreamWrapper.of(responseStreamFactory.openStream(firstReq, makeHeaders()));
+            this.nextWrapper = nextWrapper;
+            this.intTicket = intTicket;
+            this.headers = headers;
+        }
+
+        @Override
+        public void send(T payload) {
+            if (responseStream == null) {
+                responseStream = responseStreamFactory.apply(payload);
+                pending.forEach((p0, p1, p2) -> {
+                    p0.apply(responseStream);
+                    return null;
+                });
+                pending.length = 0;
+            } else {
+                //TODO handle failure of this call
```

Comment:
should this (and the one below) get the deep-retry issue number (assuming that we wouldn't be fixing all of these as part of reconnect)?